_posts/2021-09-15-drain-cleaner.md

```diff
@@ -0,0 +1,190 @@
+Operator to move pods instead of Kubernetes itself. The advantage of this approach is that the Cluster Operator makes 
+sure that no partition becomes under-replicated during node draining.
+
+As you can see from the following image, when a new Kubernetes request arrives at the API server endpoint, a [Validating 
+Admission Controller](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers) is 
+invoked after the schema validation phase and it calls every registered webhook, which implement the decision logic. 
+This can be useful to enforce custom policies or simply be notified about specific events.
+
+![admission controllers](/assets/images/posts/2021-09-15-admission-controllers.png)
+
+Some admission controllers are actually built-in (i.e. LimitRange, NamespaceLifecycle). Validating webhooks are safe, as 
+they cannot change the incoming request and they always see the final version that would be persisted to etcd. Note that 
+it may be disabled in your Kubernetes cluster for security reasons, so this is the first thing to verify.
+
+In our case, we want the Drain Cleaner to be notified of all pod eviction requests. The application will then filter out 
+all pods that are not part of our stateful sets. One important note is that Kafka CR must be configured with 
+PodDisruptionBudget's `maxUnavailable: 0` in both Kafka and Zookeeper components. This is required to block 
+Kubernetes draining and let the combined work of Drain Cleaner and Cluster Operator to handle this. 
+
+This is how you can apply this configuration with a single patch command (no rolling update is needed):
+
+```sh
+kubectl patch kafka $CLUSTER_NAME --type json -p '[
+  {
+    "op": "add",
+    "path": "/spec/zookeeper/template",
+    "value": {
+      "podDisruptionBudget": {
+        "maxUnavailable": 0
+      }
+    }
+  },
+  {
+    "op": "add",
+    "path": "/spec/kafka/template",
+    "value": {
+      "podDisruptionBudget": {
+        "maxUnavailable": 0
+      }
+    }
+  },
+]'
+```
+
+## Validating webhook configuration
+
+To register the Drain Cleaner's endpoint at the Kubernetes control-plane level, you need to create a 
+ValidatingWebhookConfiguration resource. This is a simple REST endpoint, exposed under `/drainer` URL path. When an 
+incoming API request matches one of the specified operations, groups, versions, resources, and scope for any of defined 
+rules, the request is sent to that endpoint.
+
+```yaml
+    rules:
+      - apiGroups:   [""]
+        apiVersions: ["v1"]
+        operations:  ["CREATE"]
+        resources:   ["pods/eviction"]
+        scope:       "Namespaced"
+    clientConfig:
+      service:
+        namespace: "strimzi-drain-cleaner"
+        name: "strimzi-drain-cleaner"
+        path: /drainer
+        port: 443
+      caBundle: $CA_BUNDLE
+```
+
+Here I used a placeholder for the PEM encoded CA bundle, which is required to validate the webhook's certificate. You 
```

Comment:
It is good to give the details how to do it. But maybe we could also add some mention how Cert Manager or OpenShift can help with this by generating and injecting the certs and point to the installation files?